fpga_interchange/lookahead.cc

```diff
@@ -0,0 +1,1195 @@
+        all_costs_mutex.lock();
+
+        if (ctx->verbose) {
+            log_info("Expanded all wires in type %s, merging data back\n", IdString(type_data.name).c_str(ctx));
+            log_info("Testing %zu wires, saw %zu types, deferred %zu types\n", type_data.wire_data.size(),
+                     explored->size(), deferred->size());
+        }
+
+        // Copy cheapest explored paths back to all_tiles_storage.
+        for (const auto &type_pair : storage->storage) {
+            auto &type_pair_data = all_tiles_storage.storage[type_pair.first];
+            for (const auto &delta_pair : type_pair.second) {
+                // See if this dx/dy already has data.
+                auto result = type_pair_data.emplace(delta_pair.first, delta_pair.second);
+                if (!result.second) {
+                    // This was already in the map, check if this new result is
+                    // better
+                    if (delta_pair.second < result.first->second) {
+                        result.first->second = delta_pair.second;
+                    }
+                }
+            }
+        }
+
+        // Update explored and deferred sets.
+        for (auto &key : *explored) {
+            types_explored.emplace(key);
+        }
+        for (auto &key : *deferred) {
+            types_deferred.emplace(key);
+        }
+
+        NPNR_ASSERT(tiles_left.erase(tile_type));
+
+        if (ctx->verbose) {
+            log_info("Done merging data from type %s, %zu tiles left\n", IdString(type_data.name).c_str(ctx),
+                     tiles_left.size());
+        }
+
+        all_costs_mutex.unlock();
+    });
+#else
+    }
+#endif
+
+    // Check to see if deferred wire types were expanded.  If they were not
+    // expanded, expand them now.  If they were expanded, copy_types is
+    // populated with the wire types that can just copy the relevant data from
+    // another wire type.
+#ifdef NEXTPNR_USE_TBB
+    FlatWireMap<PipAndCost> best_path(ctx);
+#endif
+    for (TypeWireId wire_type : types_deferred) {
+        auto &type_data = ctx->chip_info->tile_types[wire_type.type];
+        auto &tile_sampler = tiles_of_type[wire_type.type];
+        auto &wire_data = type_data.wire_data[wire_type.index];
+
+        if (ctx->verbose) {
+            log_info("Expanding deferred wire %s in type %s (seen %zu types)\n", IdString(wire_data.name).c_str(ctx),
+                     IdString(type_data.name).c_str(ctx), types_explored.size());
+        }
+
+        expand_deferred_routing_graph(ctx, rng, tile_sampler, wire_type, &types_explored, &all_tiles_storage,
+                                      &best_path);
+    }
+
+    auto end = std::chrono::high_resolution_clock::now();
+    if (ctx->verbose) {
+        log_info("Done with expansion, dt %02fs\n", std::chrono::duration<float>(end - start).count());
+    }
+
+    FILE *lookahead_data = fopen("lookahead.csv", "w");
```

Comment:
It looks like this is unconditionally writing a file in the current directory? this should probably be behind some kind of conditional

Reply:
Oh, this can be deleted.  This was a debugging logic I left in to visualize the output.

Reply:
I've moved this to a function, but disabled it for now.